### ISSUE TITLE
BUG: CheckFileExists/CheckFileContent performance issues 

### ISSUE THREAD

**Issue by azeemshaikh38** [ossf/scorecard#578]
The `CheckFileExists` and `CheckFileContent` functions are used by multiple checks.  These fns seem to be really slowing down the Scorecard run though. On some runs these checks can take >1hr to run.

![image](https://user-images.githubusercontent.com/4825078/121938037-2fc3fe80-cd00-11eb-9278-3d4efff0b67d.png)

The CPU profile confirms that the TAR decompression is the real culprit here: 
![image](https://user-images.githubusercontent.com/4825078/121938600-dc05e500-cd00-11eb-882b-e92d33c3f327.png)

We should update these fns (atleast the `CheckFileExists` fn) to use this API endpoint instead - https://api.github.com/repos/ossf/scorecard/git/trees/main?recursive=1. We just need to get the default branch of a repo and then make this call to get a list of all filepaths. More context: https://stackoverflow.com/questions/25022016/get-all-file-names-from-a-github-repo-through-the-github-api

@laurentsimon @naveensrinivasan do either of you have cycles to look into this? We also need a similar solution for `CheckFileContents` too.

**Comment by inferno-chromium** [comment]
if we shift to endpoint, that will work only for github repos, how about other repos if we want to support (gitlab, etc?) or even just directories as input (git-on-borg case). do these checks share the archive or keep redownloading+unpacking it again and again (that could be a nice optimization to do it once per repo). another thing could be cloning the repo in a shared place (e..g cloud filestore) and just updating it from time to time (start of cron or something ?)

**Comment by azeemshaikh38** [comment]
> if we shift to endpoint, that will work only for github repos

Not really, it would be easier to support other usecases with this endpoint in comparison to what we currently have. In terms of a generic `RepoInterface` - today we expect clients to support an equivalent of `GetRepoTarball` call. Instead, with the new endpoint we would expect clients to support `ListFiles` and `GetFileContent` call which is far more generic (especially in a local directories usecase).

> do these checks share the archive or keep redownloading+unpacking it again and again (that could be a nice optimization to do it once per repo)

As of now, they don't share the archive and the unpacking happens repeatedly. We definitely need to optimize this too.

**Comment by inferno-chromium** [comment]
> > if we shift to endpoint, that will work only for github repos
> 
> Not really, it would be easier to support other usecases with this endpoint in comparison to what we currently have. In terms of a generic `RepoInterface` - today we expect clients to support an equivalent of `GetRepoTarball` call. Instead, with the new endpoint we would expect clients to support `ListFiles` and `GetFileContent` call which is far more generic (especially in a local directories usecase).

Makes sense now, so more like an interface! And for github it is more optimized. just api quota is a pain, we can knock it off completely if we sync the repo locally. maybe good to discuss in today's meeting.

> 
> > do these checks share the archive or keep redownloading+unpacking it again and again (that could be a nice optimization to do it once per repo)
> 
> As of now, they don't share the archive and the unpacking happens repeatedly. We definitely need to optimize this too.

+1

**Comment by azeemshaikh38** [comment]
@laurentsimon assigning this to you for now.

**Comment by laurentsimon** [comment]
awesome! Sorry I got distracted yesterday and did not get the chance to respond on the bug.

**Comment by azeemshaikh38** [comment]
Did some more digging and found something interesting. Looks like there is some thread contention issues going on with the threads that use the `tar/gzip` readers, i.e checks which call `CheckIfFilesExist` and `CheckFilesContent`:

![image](https://user-images.githubusercontent.com/4825078/122331411-a07a3f00-cee9-11eb-83c5-db9ef94e6785.png)

Diagram:

![image](https://user-images.githubusercontent.com/4825078/122331511-ce5f8380-cee9-11eb-9982-c91949b163d9.png)

Basically, for some reason all these threads seem to think the underlying data is shared (which is true, not sure how this gets implied in the Go program though) and end up contending with each other causing the latency. 

Removing the `tar/gzip` will help, but I'm curious how this contention got introduced in the code and if we can fix that instead to help with the latency issues.

**Comment by oliverchang** [comment]
Hmm, I'm unsure how to interpret these graphs/stacks. What gives the indication that it's waiting on the same locks/condition vars (rather than e.g. waiting for the network)? 

From memory the tar format doesn't have some kind of central directory for all the files it contains. This means we have to seek through the whole archive to get a list of files, which might be very slow based on how we're doing it today (tar Reader on top of a http Response)?

**Comment by azeemshaikh38** [comment]
> rather than e.g. waiting for the network

I think you are right. The Golang page for `http.Response.Body` reads - `The response body is streamed on demand as the Body field is read.` So, looks like we leave the TCP connection open as we make the `tar.Next()` calls to stream in more data. Which explains the below graph in which higher latency periods correspond to higher network input:

![image](https://user-images.githubusercontent.com/4825078/122443226-38147780-cf54-11eb-9ca8-934637a3acaf.png)

Although, what's not clear is why this would block for ~1hr but then the latency disappears when you restart the job.

Anyways, this is a good candidate for trying out [`RepoInterface`](#462). We can download the tar once and the checks can share this which will reduce network ingress bytes and also remove the streaming TCP calls.

@laurentsimon assigning this to myself, hope thats ok.

**Comment by azeemshaikh38** [comment]
Improvement in network bytes received using `RepoClient` interface:

![image](https://user-images.githubusercontent.com/4825078/122480271-63fa2200-cf81-11eb-8093-36264539871a.png)

Closing this issue, but will continue monitoring to be sure.

**Comment by oliverchang** [comment]
Very nice!!
